quaternions.scad

```diff
@@ -214,65 +317,82 @@ function Q_Dist(q1, q2) = norm(q2-q1);
 //   for (q = Q_Slerp(a, b, [0:0.1:1]))
 //       Qrot(q) right(80) cube([10,10,1]);
 //   #sphere(r=80);
-function Q_Slerp(q1, q2, u) =
-    assert(is_num(u) || is_num(u[0]))
-    !is_num(u)? [for (uu=u) Q_Slerp(q1,q2,uu)] :
-    let(
-        q1 = Q_Normalize(q1),
-        q2 = Q_Normalize(q2),
-        dot = Q_Dot(q1, q2)
-    ) let(
-        q2 = dot<0? Q_Neg(q2) : q2,
-        dot = dot<0? -dot : dot
-    ) (dot>0.9995)? Q_Normalize(q1 + (u * (q2-q1))) :
+function Q_Slerp(q1, q2, u, _dot) =
+  is_undef(_dot)
+  ? assert(is_num(u) || is_range(u) || is_num(0*u*u), "Invalid interpolation coefficient(s)")
+    assert(Q_is_quat(q1) && Q_is_quat(q2), "Invalid quaternion(s)" )
     let(
-        dot = constrain(dot,-1,1),
-        theta_0 = acos(dot),
-        theta = theta_0 * u,
-        q3 = Q_Normalize(q2 - q1*dot),
-        out = q1*cos(theta) + q3*sin(theta)
-    ) out;
+      _dot  = q1*q2,
+      q1   = q1/norm(q1),
+      q2   = _dot<0 ? -q2/norm(q2) : q2/norm(q2),
+      dot = abs(_dot)
+    )
+    ! is_num(u) 
+    ? [for (uu=u) Q_Slerp(q1, q2, uu, dot)]
+    : Q_Slerp(q1, q2, u, dot) 
+  : _dot>0.9995
+    ? _Qnorm(q1 + u*(q2-q1))
+    : let( 
+        theta  = u*acos(_dot),

```

Comment:
Can `_dot` ever be approximately -1 in this code?  I have had multiple issues with `acos(x)` failing because `x` fell outside of the -1 to 1 domain due to infinitesimal FP rounding errors.  So much so that I pretty much always do a `constrain()` on a value passed to `acos()`

Reply:
Yes, it could but `dot`, the value it has in the recursion, is greater than 0; so the test `_dot>0.9995` will act as a constrain.

Reply:
Okay, I'm satisfied that this is correct.